Document embedded explorer and public variants (#1223)
apollographql/apollo: studio-docs/source/sharing-graphs.mdx

#### 2. Execute the operation

Our `executeOperation` method looks like this:

<MultiCodeBlock>

```ts:title=explorer-listener.ts
// This is the name prefix for response messages expected by the iframe
const EXPLORER_QUERY_MUTATION_RESPONSE = "ExplorerResponse";

// Replace this with your public variant's GraphQL endpoint URL!
const GRAPHQL_ENDPOINT = "https://apollo-fullstack-tutorial.herokuapp.com/";

// Replace the value of this URL's `graphRef` query parameter
// with your public variant's graph ref
const EMBEDDABLE_EXPLORER_URL = "https://explorer.embed.apollographql.com/?graphRef=Apollo-Fullstack-Demo-o3tsz8@current";

async function executeOperation({
  operation,
  operationName,
  variables,
  headers,
  embeddedExplorerIFrame,
  operationId,
}: {
  operation: string,
  operationName?: string,
  variables?: JSONValue,
  headers?: Record<string, string>,
  embeddedExplorerIFrame?: HTMLIFrameElement,
  operationId: string
}) {
  // Send the operation to the GraphQL server via fetch
  const response = await fetch(
    GRAPHQL_ENDPOINT,
    {
      method: "POST",

>       headers: getHeadersWithContentType(headers),

Reviewer: this function will need to be defined. It is just a function that adds 'content-type': 'application/json' if it isn't already there (case insensitive)

import {EmbeddableExplorer} from 'gatsby-theme-apollo-docs';

Apollo Studio provides features that help you share your graph with both teammates and external consumers.

## Customizing a graph's homepage

When a user opens one of your graph's variants in Apollo Studio, its **homepage** is shown:

Like a typical Git repository's homepage, this page displays a Markdown-based README that you can edit with any content you like.

> **Note:** Only organization members with the [`Org Admin`, `Graph Admin`, or `Contributor` role](./org/members/) can edit a variant's README. If a variant is [protected](./org/graphs/#protected-variants-enterprise-only), `Contributor`s _cannot_ edit its README.

**Each variant has its own README.** If you edit one variant's README, those edits are _not_ applied to your graph's other variants. Among other uses, this enables you to provide different information to external consumers in a [public variant](#making-a-graph-public).

### README shortcodes

The README supports a collection of variables that you can include in its text. These variables are rendered with the correct value for the current variant.

## Public graphs

> You can make individual [variants](./org/graphs/#managing-variants) of your graph public. If you do, **anyone with the link** can view the following pages for that variant in Studio:

Reviewer 2: @StephenBarlow anyone with the link seems scary. Maybe we can modify to say _anyone with the link to the public variant_? We may also want to add that the link will have a /public/ in the URL?

People outside your organization can't view any _other_ pages for a public variant (Fields, Operations, etc.), and they can never view _any_ pages for private variants. New variants always start as private.

### Making a variant public

> **Note:** Only organization members with the [`Org Admin` or `Graph Admin` role](./org/members/) can toggle a variant's visibility.

1. Go to your graph's Settings page and open the Access tab.
2. Under the Variants section, toggle the switch for the variant you want to make public.
3. Confirm your selection.

You can toggle the switch back to make the variant private again.

## Embedding the Explorer

If you have a [public variant](#public-graphs) of your graph, you can embed the [Apollo Studio Explorer](./explorer/) in a webpage that you can then provide to your graph's consumers. This enables those consumers to test out operations from your own website.

For example, here's an embedded Explorer for an Apollo example graph:

<EmbeddableExplorer
  graphRef="Apollo-Fullstack-Demo-o3tsz8@current"
  graphEndpoint="https://apollo-fullstack-tutorial.herokuapp.com/"
  graphSubscriptionEndpoint="wss://apollo-fullstack-tutorial.herokuapp.com/graphql"

>   styles={{width: '100%', height:'450px'}}

Reviewer: ```suggestion
  styles={{width: '100%', height:'450px', border: '0px', borderRadius: '4px'}}
```

Iframes by default have an ugly grey `frameBorder` we could remove it on the react component or here

      const { operation, operationName, variables, headers } = event.data;

      // Execute the operation, providing all required details
      // (we'll define executeOperation next)
      executeOperation({
        operation: event.data.operation,
        operationName,
        variables,
        headers,
        embeddedExplorerIFrame,
        operationId,
      });
    }
  };

  // Execute our callback whenever window.postMessage is called
  window.addEventListener("message", onPostMessageReceived);
}
```

</MultiCodeBlock>

Now we can import and call `setupEmbedRelay` on any page where we'll embed the Explorer. However, we still need to define the `executeOperation` function that's called by our callback.

#### 3. Define operation execution logic

Our webpages can use the `fetch` API to send operations to our GraphQL server. To do this, we need the URL of our server's GraphQL endpoint. Substitute it in the `executeOperation` function below, where indicated.

<MultiCodeBlock>

```ts:title=setupEmbedRelay.ts
// Helper function that adds content-type: application/json
// to each request's headers if not present

> function getHeadersWithContentType(

Reviewer: Oh did you always have this here? sorry if I missed it last time!

Author: Ah no not at all! I added it thanks to your comment!

        query: operation,
        variables,
        operationName,
      }),
    }
  );

  // After the operation completes, post a response message to the
  // iframe that includes the response data
  await response.json().then((response) => {
    embeddedExplorerIFrame?.contentWindow?.postMessage(
      {
        // Include the same operation ID in the response message's name
        // so the Explorer knows which operation it's associated with
        name: `${EXPLORER_QUERY_MUTATION_RESPONSE}:${operationId}`,
        response,
      },
      EMBEDDABLE_EXPLORER_URL
    );
  });
}
```

</MultiCodeBlock>

Instead of hardcoding your server's URL as shown, you can modify `executeOperation` to take a server URL parameter if you'll be embedding Explorers for multiple graphs.

#### 4. Embed the `iframe`

We're ready to add an embedded Explorer to our page! Here's an example `App.tsx` file that uses our `setupEmbedRelay` function to wire up an embedded Explorer `iframe`:

<MultiCodeBlock>

```tsx:title=App.tsx
import React, { useEffect } from 'react';
import { EMBEDDABLE_EXPLORER_URL, setUpEmbedRelay } from './setupEmbedRelay';

export const App = ()=> {

  useEffect(() => {

>     setUpEmbedRelay();

Reviewer: If we are going to include this as a React demo, we might want to do it in a more React-y way (including cleaning up the listener on unmount). I did it this way because I thought people would want to vanilla js in the docs. Thoughts?

Context: https://apollograph.slack.com/archives/C027DK254DS/p1629937720000500?thread_ts=1629935176.086300&cid=C027DK254DS

```
  useEffect(() => {
    const onPostMessageReceived = (event:MessageEvent<{
      name?: string,
      operation?: string,+      operationName?: string,
      variables?: string,
      headers?:string
    }>) => {
      const isQueryOrMutation = event.data.name?.startsWith('ExplorerRequest:');
      const currentOperationId = event.data.name?.split(':')[1]
      if(isQueryOrMutation && event.data.operation) {
          executeOperation({
            operation: event.data.operation,
            operationName: event.data.operationName,
            variables: event.data.variables,
          }).then((response) => {
            const embeddedExplorerIFrame = document.getElementById('embedded-explorer') as HTMLIFrameElement;
            embeddedExplorerIFrame?.contentWindow?.postMessage({
              name: isQueryOrMutation ?
                `ExplorerResponse:${currentOperationId}` :
                `ExplorerSubscriptionResponse:${currentOperationId}`,
              response,
            }, EMBEDDABLE_EXPLORER_URL);
          });
      }
    }
    window.addEventListener('message', onPostMessageReceived);

    return () => window.removeEventListener('message', onPostMessageReceived);
  }, [])

```

I can also make a vanilla js demo it is just more annoying with the babel and package managers etc that create react app does for you

  }, [])

  // Provide iframe options via URL query parameters
  const explorerURL = EMBEDDABLE_EXPLORER_URL +
    '?graphRef=Apollo-Fullstack-Demo-o3tsz8@current' +
    '&docsPanelState=closed';

  return (
    <div className="App">
      <h1 className="demo-app-title">Apollo's Embedded Explorer Demo App</h1>
      <iframe id="embedded-explorer" className="embedded-explorer-iframe" title="embedded-explorer" src={explorerURL} />
    </div>
  );
}
```

</MultiCodeBlock>

> Notice that we append some query parameters to our `iframe`'s URL, most importantly `graphRef`. This tells the Explorer which [public variant's](#public-graphs) schema to download to populate the Docuentation panel and enable symbol completion.

Reviewer: ```suggestion
Notice that we append some query parameters to our `iframe`'s URL, most importantly `graphRef`. This tells the Explorer which [public variant's](#public-graphs) schema to download to populate the Documentation panel and enable symbol completion.
```

  <img alt="Edit embedded-explorer-demo" src="https://codesandbox.io/static/img/play-codesandbox.svg" />
</a>

The example includes support for subscriptions, which are omitted from this Setup section for clarity. The subscription-specific logic should be more straightforward after familiarizing yourself with the logic for queries and mutations.

### Options

When you embed the Explorer, you set options for it by including query parameters in the `src` URL of the `iframe`, like so:

```jsx:title=EmbeddedExplorer.jsx
return (
  <iframe src={`https://explorer.embed.apollographql.com/?graphRef=${myGraphRef}&docsPanelState=closed`} />
);
```

The following options are supported:

<table class="field-table api-ref">
  <thead>
    <tr>
      <th>Name /<br/>Type</th>
      <th>Description</th>
    </tr>
  </thead>

<tbody>
<tr class="required">
<td>

##### `graphRef`

`string`

</td>
<td>

**Recommended.** The graph ref for the [public variant](#public-graphs) you want to use the embedded Explorer with. Has the format `graph-name@variant-name`.

The Explorer fetches this variant's schema from Apollo Studio to populate its Documentation panel and enable symbol completion.

> If you omit this option, the Explorer still loads, but it can't provide any useful information about the graph the user is querying against.

Reviewer 2: @StephenBarlow Maybe we can make this more specific?

`If you omit this option, the Explorer still loads, but it can't provide any useful information about the graph the user is querying against as the Explorer does not know which schema to load. Providing a graph ref helps the Explorer to fetch the schema it needs to load.`

Reviewer 2: Or something more succinct but clarifying that `Explorer does not know what schema to load if you don't do this, SO DO IT!` 😛

---
title: Sharing your graph
---

import {
  ExpansionPanel,
} from 'gatsby-theme-apollo-docs/src/components/expansion-panel';

Apollo Studio provides features that help you share your graph with both teammates and external consumers.

## Making a graph public

> You can make individual [variants](./org/graphs/#managing-variants) of your graph public. If you do, people _outside_ your organization can view the following pages for that variant in Studio:

Reviewer 2: @StephenBarlow Do you think it is sufficiently clear that any unauthenticated user on the internet who has access to the public graph link will be able to open the graph schema? Not sure if it is just because I have been living in it for so long that this might make me wonder if I have more than 1 Studio org, then it is *not* just that an authenticated user of another org can access, it is any unauthenticated user.

Reviewer 2: To clarify - any unauthenticated user with the public graph link will be able to access the schema

* Home
* Schema
* Explorer

> This enables consumers of your graph to learn about your graph's schema and any special usage information (such as authentication details) that you've documented in [the variant's README](#customizing-a-graphs-homepage). They can also run operations against your graph with the Explorer.

Reviewer 2: @StephenBarlow Should we clarify that user will still need to authenticate the GraphQL endpoint?

People outside your organization can't view any _other_ pages for a public variant (Fields, Operations, etc.), and they can never view _any_ pages for private variants. New variants always start as private.

> **Note:** Only organization members with the [`Org Admin` or `Graph Admin` role](./org/members/) can set a variant's visibility.

> **To make a variant public:**

Reviewer 2: @StephenBarlow Capability to make a variant public or not is only available to org/graph admins.

1. Go to your graph's Settings page and open the Access tab.
2. Under the Variants section, toggle the switch for the variant you want to make public.
3. Confirm your selection.

You can make a public variant private by toggling the switch back.

## Embedding the Explorer

TODO

## Customizing a graph's homepage

When a user opens one of your graph's variants in Apollo Studio, its **homepage** is shown:

Like a typical Git repository's homepage, this page displays a Markdown-based README that you can edit with any content you like.

> **Note:** Only organization members with the [`Org Admin`, `Graph Admin`, or `Contributor` role](./org/members/) can edit a variant's README. If a variant is [protected](./org/graphs/#protected-variants-enterprise-only), `Contributor`s _cannot_ edit its README.

> **Each variant has its own README.** If you edit one variant's README, those edits are _not_ applied to your graph's other variants. Among other uses, this enables you to provide different information to external consumers in a [public variant](#making-a-graph-public).

Reviewer 2: TO ADD: README markdown syntax guide

Reviewer 2: @cy or @cheapsteak to provide all the supported syntax

      <th>Description</th>
    </tr>
  </thead>

<tbody>
<tr class="required">
<td>

##### `graphRef`

`string`

</td>
<td>

**Required.** The graph ref for the [public variant](#public-graphs) you want to use the embedded Explorer with. Has the format `graph-name@variant-name`.

The Explorer fetches this variant's schema from Apollo Studio to populate its Documentation panel and enable symbol completion.

If you omit this option, the Explorer _does_ still load, but it doesn't know which schema to fetch. This prevents the Explorer from providing critical features like documentation and symbol completion.

</td>
</tr>

<tr>
<td>

##### `docsPanelState`

`"open" | "closed"`

</td>
<td>

If `open`, the Explorer's Documentation panel (the left column) is initially expanded. If `closed`, the panel is initially collapsed.

The default value is `open`.

</td>
</tr>

<tr>
<td>

##### `showHeadersAndEnvVars`

`"true" | "false"`

</td>
<td>

If `true`, the embedded Explorer includes the panels for [setting request headers and environment variables](./explorer/#authentication). If `false`, those panels are not present.

The default value is `true`.

</td>
</tr>

<tr>
<td>

> 

Reviewer: ```suggestion
##### `performOperationsInEmbed`

`"true" | "false"`

</td>
<td>

If `true`, when you submit an operation through the explorer, we will use your graph ref's endpoint to perform operations, instead of post messaging the network requests to the parent page. 

The default value is `false`.

</td>
</tr>
</tbody>
</table>
```

Reviewer: This is something we are adding in! PR: https://github.com/mdg-private/studio-ui/pull/4904

Reviewer: You have to be on a public graph with a url endpoint set that that to work, though.

1. Go to your graph's Settings page and open the Access tab.
2. Under the Variants section, toggle the switch for the variant you want to make public.
3. Confirm your selection.

You can toggle the switch back to make the variant private again.

## Embedding the Explorer

If you have a [public variant](#public-graphs) of your graph, you can embed the [Apollo Studio Explorer](./explorer/) in a webpage that you can then provide to your graph's consumers. This enables those consumers to test out operations from your own website.

For example, here's an embedded Explorer for an Apollo example graph:

<EmbeddableExplorer
  graphRef="Apollo-Fullstack-Demo-o3tsz8@current"
  graphEndpoint="https://apollo-fullstack-tutorial.herokuapp.com/"
  graphSubscriptionEndpoint="wss://apollo-fullstack-tutorial.herokuapp.com/graphql"
  styles={{width: '100%', height:'450px', border: '0px', borderRadius: '4px'}}
/>

> This embedded Explorer collapses its left column by default because the full three-column layout is a little cramped on this page. You can [customize the embedded Explorer's appearance](#options) to suit the page it's on.

### Basics

> You embed the Explorer in an `iframe` that points to `https://explorer.embed.apollographql.com`. You add a `graphRef` query parameter to this URL to indicate which [public variant](#public-graphs) the Explorer should use:

Reviewer: ```suggestion
You embed the Explorer in an `iframe` that points to `https://explorer.embed.apollographql.com`. You add a `graphRef` query parameter to this URL to indicate which [public variant](#public-graphs) the Explorer should use, and a `postMessageOperations` query parameter to indicate whether you want to make network requests from the embedded Explorer or from your web page:
```

Reviewer: Not sure if this is where we want it, so just leaving suggestions

```jsx
<iframe src="https://explorer.embed.apollographql.com/?graphRef=my-graph@my-variant" />
```

The embedded Explorer uses this `graphRef` to fetch your public variant's schema and its GraphQL endpoint URL from Apollo Studio.

#### Handling CORS

> By default, the embedded Explorer sends user-executed operations directly to your variant's GraphQL endpoint. Because the Explorer is embedded in an `iframe`, the origin of those network requests is `https://explorer.embed.apollographql.com`.

Reviewer: ```suggestion
The embedded Explorer can either send user-executed operations directly to your variant's GraphQL endpoint, or send them to your webpage via the `postMessage` API. If you pass `postMessageOperations=true`, because the Explorer is embedded in an `iframe`, the origin of those network requests is `https://explorer.embed.apollographql.com`.
```

### Supported Markdown

The README supports all [basic Markdown syntax](https://www.markdownguide.org/cheat-sheet/#basic-syntax). Most [extended syntax](https://www.markdownguide.org/cheat-sheet/#extended-syntax) is _not_ fully supported, except for code blocks.

You can define a GraphQL code block with the following syntax:

    ```graphql
    query MyQuery {
      users {
        id
        name
      }
    }
    ```

This renders the code block with buttons you can use to copy an operation or open it in the Explorer:

<img class="screenshot" alt="GraphQL codeblock in README" src="./img/readme-codeblock.jpg" width="300" />

### README shortcodes

The README supports a set of shortcodes you can use to populate it with data that's specific to your variant, such as the graph's name and the variant's graph ref:

```md
# Overview of {{ graph.name }}

Graph ref: {{ graph.ref }}
```

For a list of all supported shortcodes, click **Graph shortcodes** in the bottom-left corner of the README editor.

> ## Public graph access

Reviewer 2: `Public access `for/on graphs?

**If you need requests to originate from your webpage**, additional setup is required. See [Using `postMessage`](#using-postmessage).

### Options

When you embed the Explorer, you set options for it by including query parameters in the `src` URL of the `iframe`, like so:

```jsx
<iframe src="https://explorer.embed.apollographql.com/?graphRef=my-graph@my-variant&theme=light" />
```

The following options are supported:

<table class="field-table api-ref">
  <thead>
    <tr>
      <th>Name /<br/>Type</th>
      <th>Description</th>
    </tr>
  </thead>

<tbody>
<tr class="required">
<td>

##### `graphRef`

`string`

</td>
<td>

**Required.** The graph ref for the [public variant](#public-graph-access) you want to use the embedded Explorer with. Has the format `graph-name@variant-name`.

The Explorer fetches this variant's schema from Apollo Studio to populate its Documentation panel and enable symbol completion.

If you omit this option, the Explorer _does_ still load, but it doesn't know which schema to fetch. This prevents the Explorer from providing critical features like documentation and symbol completion.

</td>
</tr>

> 

Reviewer: ```suggestion
<tr>
<td>

##### `explorerURLState`

URI encoded operation, variables, headers and search query. You can get a value for `explorerURLState` by going to your public variant, populating the operation, variables, headers and searchQuery that you want to copy, click on the share icon at the top right of the editor pane, and copy the explorerURLState query param.

</td>
<td>

If `open`, the Explorer's Documentation panel (the left column) is initially expanded. If `closed`, the panel is initially collapsed.

The default value is `open`.

</td>
</tr>

<tr>
<td>
```

## Embedding the Explorer

If you have a [public variant](#making-a-variant-public) of your graph, you can embed the [Apollo Studio Explorer](./explorer/) in a webpage that you can then provide to your graph's consumers. This enables those consumers to test out operations from your own website.

For example, here's an embedded Explorer for an Apollo example graph. Try it out!

<EmbeddableExplorer
  graphRef="Apollo-Fullstack-Demo-o3tsz8@current"
  graphEndpoint="https://apollo-fullstack-tutorial.herokuapp.com/"
  graphSubscriptionEndpoint="wss://apollo-fullstack-tutorial.herokuapp.com/graphql"
  defaultOperation={`query GetLaunches {
  launches {
    launches {
      id
      site
      rocket {
        id
        name
      }
    }
  }
}`}
  styles={{width: '100%', height:'450px', border: '0px', borderRadius: '4px'}}
/>

> This embedded Explorer collapses its left column by default because the full three-column layout is a little cramped on this page. You can [customize the embedded Explorer's appearance](#options) to suit the page it's on.

### Basic syntax

You embed the Explorer in an `iframe` that points to `https://explorer.embed.apollographql.com`:

```jsx

> <iframe src="https://explorer.embed.apollographql.com/?graphRef=my-graph-id@my-variant&postMessageOperations=false" />

Reviewer: ```suggestion
<iframe src="https://explorer.embed.apollographql.com/?graphRef=my-graph-id@my-variant&sendRequestsFrom=embed" />
```

```

As shown, the `iframe`'s `src` URL requires the following query parameters:

* `graphRef` - This indicates which [public variant's](#public-access-for-graphs) schema the Explorer should download from Apollo Studio (e.g., `my-graph-id@my-variant`).

> * `postMessageOperations` - This indicates whether the `iframe` sends its own network requests to your GraphQL endpoint.

Reviewer: ```suggestion
* `sendRequestsFrom` - This indicates whether the `iframe` sends its own network requests to your GraphQL endpoint.
```

    * You set this according to how you [handle CORS](#handling-cors).

### Handling CORS

Your GraphQL endpoint needs to be able to accept requests from your embedded Explorer. Because the Explorer is embedded in an `iframe`, its requests originate from `https://explorer.embed.apollographql.com`.

There are two supported methods for handling this: [adding a CORS domain](#add-a-cors-domain-recommended) or [routing requests through your webpage](#route-requests-through-your-webpage-advanced).

#### Add a CORS domain (recommended)

We recommend that you update your GraphQL endpoint's CORS configuration (such as with the [`cors` `ApolloServer` option](https://www.apollographql.com/docs/apollo-server/api/apollo-server/#cors)) to accept requests from `https://explorer.embed.apollographql.com`.

> > If you use this method, set the `postMessageOperations` query parameter to `false`. Also make sure you've set your public variant's endpoint URL in Apollo Studio.

Reviewer: ```suggestion
> If you use this method, set the `sendRequestsFrom` query parameter to `embed`. Also make sure you've set your public variant's endpoint URL in Apollo Studio.
```

After you add the Explorer's origin to your endpoint's CORS configuration, you're all set! You can now embed the `iframe` in any webpage and set its dimensions as you like. Also check out [additional supported options](#options).

#### Route requests through your webpage (advanced)

Some endpoints _can't_ receive requests from the Explorer's origin due to firewall rules or organizational policies. In these cases, you can use the the `postMessage` API to route Explorer requests through your embedding webpage's domain.

> > If you use this method, set the `postMessageOperations` query parameter to `true`.

Reviewer: ```suggestion
> If you use this method, set the `sendRequestsFrom` query parameter to `parent`.
```

If you need to use this advanced method, proceed to the additional setup in [Using `postMessage`](#using-postmessage-advanced).

### Options

When you embed the Explorer, you set options for it by including query parameters in the `src` URL of the `iframe`, like so:

```jsx

> <iframe src="https://explorer.embed.apollographql.com/?graphRef=my-graph-id@my-variant&postMessageOperations=false" />

Reviewer: ```suggestion
<iframe src="https://explorer.embed.apollographql.com/?graphRef=my-graph-id@my-variant&sendRequestsFrom=embed" />
```

```

The following options are supported:

<table class="field-table api-ref">
  <thead>
    <tr>
      <th>Name /<br/>Type</th>
      <th>Description</th>
    </tr>
  </thead>

<tbody>
<tr class="required">
<td>

##### `graphRef`

`string`

</td>
<td>

**Required.** The graph ref for the [public variant](#public-access-for-graphs) you want to use the embedded Explorer with. Has the format `graph-id@variant-name`.

The Explorer fetches this variant's schema from Apollo Studio to populate its Documentation panel and enable symbol completion.

If you omit this option, the Explorer _does_ still load, but it doesn't know which schema to fetch. This prevents the Explorer from providing critical features like documentation and symbol completion.

</td>
</tr>

<tr class="required">
<td>

> ##### `postMessageOperations`

Reviewer: ```suggestion
##### `sendRequestsFrom`
```

`"true" | "false"`

</td>
<td>

> **Required.** If `true`, the embedded Explorer uses the `postMessage` API to notify your webpage when a user initiates a GraphQL operation. Your webpage is then responsible for transmitting the operation to your GraphQL server.

Reviewer: ```suggestion
**Required.** If set to `parent`, the embedded Explorer uses the `postMessage` API to notify your webpage when a user initiates a GraphQL operation. Your webpage is then responsible for transmitting the operation to your GraphQL server.
```

> If `false`, the embedded Explorer transmits operations directly to your GraphQL server from the origin `https://explorer.embed.apollographql.com`.

Reviewer: ```suggestion
If set to `embed`, the embedded Explorer transmits operations directly to your GraphQL server from the origin `https://explorer.embed.apollographql.com`.
```

See [Handling CORS](#handling-cors).

</td>
</tr>

<tr>
<td>

##### `document`

`string`

</td>
<td>

A URI-encoded operation to populate in the Explorer's editor on load.

If you omit this, the Explorer initially loads an example query from your schema. It then saves the user's last operation in local storage for future page loads.

Example:

```jsx
?document=${window.encodeURIComponent(`
  query ExampleQuery {
    id
  }
`)}
```

</td>
</tr>

<tr>
<td>

##### `variables`

`string`

</td>
<td>

A URI-encoded, serialized object containing initial variable values to populate in the Explorer on load.

If provided, these variables should apply to the initial query you provide for [`document`](#document).

Example:

```jsx
?variables=${window.encodeURIComponent(
  JSON.stringify({
    userID: "abc123"
  })
)}
```

</td>
</tr>

<tr>
<td>

##### `headers`

`string`

</td>
<td>

A URI-encoded, serialized object containing initial HTTP header values to populate in the Explorer on load.

Example:

```jsx
?headers=${window.encodeURIComponent(
  JSON.stringify({
    authorization: "Bearer abc123"
  })
)}
```

</td>
</tr>

<tr>
<td>

> 

Reviewer: ```suggestion
<tr>
<td>

##### `shouldPersistState`

`boolean`

</td>
<td>

If `true`, we will use local storage and persist user state between sessions based on `graphRef` & hostname embedding the embedded Explorer. 

If `false`, each load of the embedded Explorer will be a fresh state for the user.

</td>
</tr>
```

        // so the Explorer knows which operation it's associated with
        name: EXPLORER_QUERY_MUTATION_RESPONSE,
        operationId,
        response,
      },
      EMBEDDABLE_EXPLORER_URL
    );
  });
}
```

</MultiCodeBlock>

Instead of hardcoding your server's URL as shown, you can modify `executeOperation` to take a server URL parameter if you'll be embedding Explorers for multiple graphs.

#### 4. Embed the `iframe`

We're ready to add an embedded Explorer to our page! Below is an example `App.tsx` file that uses our `setupEmbedRelay` function to wire up an embedded Explorer `iframe`.

> This example uses React, but you can use any view layer you like. If you _do_ use React, you should also modify your code to remove your `message` event listener when your component unmounts.

<MultiCodeBlock>

```tsx:title=App.tsx
import React, { useEffect } from 'react';
import { EMBEDDABLE_EXPLORER_URL, setUpEmbedRelay } from './setupEmbedRelay';

export const App = ()=> {

  useEffect(() => {
    setUpEmbedRelay();
  }, [])

  // Provide iframe options via URL query parameters
  const explorerURL = EMBEDDABLE_EXPLORER_URL +
    '?graphRef=Apollo-Fullstack-Demo-o3tsz8@current' +

>     '&postMessageOperations=true';

Reviewer: ```suggestion
    '&sendRequestFrom=parent';
```

For example, here's an embedded Explorer for an Apollo example graph:

<EmbeddableExplorer
  graphRef="Apollo-Fullstack-Demo-o3tsz8@current"
  graphEndpoint="https://apollo-fullstack-tutorial.herokuapp.com/"
  graphSubscriptionEndpoint="wss://apollo-fullstack-tutorial.herokuapp.com/graphql"
  styles={{width: '100%', height:'450px', border: '0px', borderRadius: '4px'}}
/>

> This embedded Explorer collapses its left column by default because the full three-column layout is a little cramped on this page. You can [customize the embedded Explorer's appearance](#options) to suit the page it's on.

### Basic setup

You embed the Explorer in an `iframe` that points to `https://explorer.embed.apollographql.com`:

```jsx
<iframe src="https://explorer.embed.apollographql.com/?graphRef=my-graph@my-variant&postMessageOperations=false" />
```

As shown, the `iframe`'s `src` URL requires two query parameters:

* `graphRef`, which indicates which [public variant's](#public-access-for-graphs) schema the Explorer should download from Apollo Studio (e.g., `my-graph@my-variant`)

> * `postMessageOperations`, which indicates whether the `iframe` is responsible for sending its own network requests to your GraphQL server (this depends on how you [handle CORS](#handling-cors))

Reviewer: ```suggestion
* `postMessageOperations`, which indicates whether the `iframe` is responsible for sending its own network requests to your GraphQL server (this depends on how you [handle CORS](#handling-cors)), or using the endpoint urls you have set on your public variant in studio.
```

#### Handling CORS

The embedded Explorer can transmit GraphQL operations directly to your GraphQL server. Because the Explorer is embedded in an `iframe`, the origin of its network requests is `https://explorer.embed.apollographql.com`.

Because of this, you need to do one of the following:

* Update your GraphQL server's CORS configuration (such as with the [`cors` `ApolloServer` option](https://www.apollographql.com/docs/apollo-server/api/apollo-server/#cors)) to accept requests from `https://explorer.embed.apollographql.com`
    * This method is more straightforward if your organization allows it.
* Use the `postMessage` API to send requests from your webpage's domain _instead of_ the `iframe` (advanced)

> **If you're sending requests from the `iframe`**, set the `iframe` URL's `postMessageOperations` parameter to `false`. You can now add the `iframe` to any webpage and set its dimensions as you like. Also check out [additional supported options](#options).

Reviewer: ```suggestion
**If you're sending requests from the `iframe`**, make sure you have an endpoint url set on your public variant on studio, and set the `iframe` URL's `postMessageOperations` parameter to `false`. You can now add the `iframe` to any webpage and set its dimensions as you like. Also check out [additional supported options](#options).
```

**If you're sending requests from your webpage's domain**, set the `iframe` URL's `postMessageOperations` parameter to `true`. Then proceed to the additional setup in [Using `postMessage`](#using-postmessage-advanced).

### Options

When you embed the Explorer, you set options for it by including query parameters in the `src` URL of the `iframe`, like so:

```jsx
<iframe src="https://explorer.embed.apollographql.com/?graphRef=my-graph@my-variant&postMessageOperations=false" />
```

The following options are supported:

<table class="field-table api-ref">
  <thead>
    <tr>
      <th>Name /<br/>Type</th>
      <th>Description</th>
    </tr>
  </thead>

<tbody>
<tr class="required">
<td>

##### `graphRef`

`string`

</td>
<td>

> **Required.** The graph ref for the [public variant](#public-access-for-graphs) you want to use the embedded Explorer with. Has the format `graph-name@variant-name`.

Reviewer: ```suggestion
**Required.** The graph ref for the [public variant](#public-access-for-graphs) you want to use the embedded Explorer with. Has the format `graph-id@variant-name`.
```

Reviewer: graph name & id are different in studio, confusing 😞

<MultiCodeBlock>

```ts:title=setupEmbedRelay.ts
// URL for any embedded Explorer iframe
export const EMBEDDABLE_EXPLORER_URL =
  "https://explorer.embed.apollographql.com";

// Message types for queries and mutations
const EXPLORER_QUERY_MUTATION_REQUEST = "ExplorerRequest";
const EXPLORER_QUERY_MUTATION_RESPONSE = "ExplorerResponse";

// Message types for subscriptions (not covered in this setup)
const EXPLORER_SUBSCRIPTION_REQUEST = "ExplorerSubscriptionRequest";
const EXPLORER_SUBSCRIPTION_RESPONSE = "ExplorerSubscriptionResponse";
const SUBSCRIPTION_TERMINATION = "ExplorerSubscriptionTermination";
```

</MultiCodeBlock>

#### 2. Set up communication

Next, let's define a function that sets up communication between a webpage and the embedded Explorer's `iframe`. This function defines a callback that will execute whenever `window.postMessage` is called by the `iframe`.

<MultiCodeBlock>

```ts:title=setupEmbedRelay.ts
export function setUpEmbedRelay() {

  // Callback definition
  const onPostMessageReceived = (
    event: MessageEvent<{
      name?: string;
      operation?: string;
      operationId?: string;
      operationName?: string;

>       variables?: string;

Reviewer: ```suggestion
      variables?: Record<string, string>;
```

Lets just make this an object everywhere in the docs. I will add suggestions for all the places it is used

      headers?: Record<string, string>;
    }>
  ) => {

    // Obtain the iframe element by any applicable logic for your page.
    // This obtains an element with ID `embedded-explorer`.
    const embeddedExplorerIFrame =
      (document.getElementById("embedded-explorer") as HTMLIFrameElement) ??
      undefined;

    // Check to see if the posted message indicates that the user is
    // executing a query or mutation in the Explorer
    const isQueryOrMutation =
      "name" in event.data &&
      event.data.name === EXPLORER_QUERY_MUTATION_REQUEST;

    // If the user is executing a query or mutation...
    if (
      isQueryOrMutation &&
      event.data.name &&
      event.data.operation &&
      event.data.operationId
    ) {
      // Extract the operation details from the event.data object
      const { operation, operationId, operationName, variables, headers } = event.data;

      // Execute the operation, providing all required details
      // (we'll define executeOperation next)
      executeOperation({

>         operation: event.data.operation,

Reviewer: ```suggestion
        operation,
```

Reviewer: unnecessary / typo from me

Now we can import and call `setupEmbedRelay` on any page where we'll embed the Explorer. However, we still need to define the `executeOperation` function that's called by our callback.

#### 3. Define operation execution logic

Our webpages can use the `fetch` API to send operations to our GraphQL server. To do this, we need the URL of our server's GraphQL endpoint. Substitute it in the `executeOperation` function below, where indicated.

<MultiCodeBlock>

```ts:title=setupEmbedRelay.ts
// Helper function that adds content-type: application/json
// to each request's headers if not present
function getHeadersWithContentType(
  headers: Record<string, string> | undefined
) {
  const headersWithContentType = headers ?? {};
  if (
    Object.keys(headersWithContentType).every(
      (key) => key.toLowerCase() !== "content-type"
    )
  ) {
    headersWithContentType["content-type"] = "application/json";
  }
  return headersWithContentType;
}

// Function for executing operations
async function executeOperation({
  operation,
  operationId,
  operationName,
  variables,
  headers,
  embeddedExplorerIFrame,
}: {
  operation: string;
  operationId: string;
  operationName?: string;

>   variables?: JSONValue;

Reviewer: ```suggestion
  variables?: Record<string, string>;
```

    "https://apollo-fullstack-tutorial.herokuapp.com/",
    {
      method: "POST",
      headers: getHeadersWithContentType(headers),
      body: JSON.stringify({
        query: operation,
        variables,
        operationName,
      }),
    }
  );

  // After the operation completes, post a response message to the
  // iframe that includes the response data
  await response.json().then((response) => {
    embeddedExplorerIFrame?.contentWindow?.postMessage(
      {
        // Include the same operation ID in the response message's name
        // so the Explorer knows which operation it's associated with
        name: EXPLORER_QUERY_MUTATION_RESPONSE,
        operationId,
        response,
      },
      EMBEDDABLE_EXPLORER_URL
    );
  });
}
```

</MultiCodeBlock>

Instead of hardcoding your server's URL as shown, you can modify `executeOperation` to take a server URL parameter if you'll be embedding Explorers for multiple graphs.

#### 4. Embed the `iframe`

We're ready to add an embedded Explorer to our page! Below is an example `App.tsx` file that uses our `setupEmbedRelay` function to wire up an embedded Explorer `iframe`.

> > This example uses React, but you can use any view layer you like. If you _do_ use React, you should also modify your code to remove your `message` event listener when your component unmounts.

Reviewer: 🙏 thank you for this!

The following operation details are included in the associated `event.data` object:

* `operation` (string)
* `operationId` (string)
* `operationName` (string)
* `variables` (object)
* `headers` (object)

</td>
</tr>

<tr>
<td>

##### `ExplorerResponse`

</td>
<td>

Posted by your page to notify the `iframe` of a query or mutation result. The message should be the operation's JSON response object.

</td>
</tr>

<tr>
<td>

##### `ExplorerSubscriptionRequest`

</td>
<td>

Posted by the `iframe` to notify your page that the user wants to execute a subscription operation.

The same operation details are included in the `event.data` object that are included in [`ExplorerRequest` messages](#explorerrequest).

</td>
</tr>

<tr>
<td>

##### `ExplorerSubscriptionResponse`

</td>
<td>

Posted by your page to notify the `iframe` of a new subscription result. The message should be the subscription's new JSON response object.

</td>
</tr>

<tr>
<td>

##### `ExplorerSubscriptionTermination`

</td>
<td>

Posted by the `iframe` to notify your page that the user wants to terminate an active subscription operation.

</td>
</tr>

<tr>
<td>

##### `SetOperation`

</td>
<td>

> Posted by your page to provide the `iframe` with an operation string to populate the Explorer's editor. You might want to use this to populate a default operation on page load.

Reviewer: Now that we have tabs, this will currently replace the last open tab. I am changing it to open a new tab, but not sure if the copy needs to change for tabs. (I think its fine?)
